### DISCUSSION TITLE
Cannot load project due to tools.jar

### DISCUSSION THREAD

**Discussion by demlik** [eclipse-m2e/m2e-core#1037] [Q&A]
Hi, I am trying to import a project which for one reason or another has a system dependency to JDK tools.jar.

All versions of Eclipse which now require Java11+ seem to have issue resolving this dependency even when JDK8 is added to installed JRE section with tools.jar added as an external dependency.

Reverting to a version of Eclipse which can be run under JDK8 (via -vm) works perfectly fine.

Maybe m2e is using the internal JREs home directory instead of the specified JREs home directory…

Is there a way to resolve this issue?

**Comment by glhez** [comment]
I have a similar issue : trying to reference `tools.jar` using these as reference: https://stackoverflow.com/questions/3080437/jdk-tools-jar-as-maven-dependency

The problem is that to reference the `tools.jar`, you would often needs to do  `${java.home}/../lib/tools.jar` which is a path only working in Java <=8. 

Since Eclipse and Lemminx is running Java 17, the `${java.home}` will resolve to that JDK.

Not sure what can be done, but in Eclipse, you can add tools.jar in the JRE. 
And perhaps use profile machinery for Maven builds: https://github.com/olivergondza/maven-jdk-tools-wrapper

Note that if you were to use toolchains to build in Java 8 with Maven running with Java 17, you would have the same problem.

Another alternative is to explicitly use / declare a java8.home.

**Reply by laeubi** [reply]
The best option seems to get rid of `tools.jar`, so what are you really need it only for compilation as indicated in the linked issue? Then [maven-jdk-tools-wrapper](https://github.com/olivergondza/maven-jdk-tools-wrapper) seems the best option + migrating your code-base to a more modern java version.

**Reply by mickaelistria** [reply]
The proper solution would be that you invest a bit in stopping to require tools.jar, so your code isn't stuck to Java 8. Java 8 isn't really supported any longer; and as you noticed, it's now getting harder and harder to get Eclipse bits properly supporting it (and there are no real chances this gets better; unless you're ready to throw tons of money to some Eclipse developers to take care of such compatibility).

However, while I don't think there is an easy or even existing solution, but some pieces are there:
First you should use the `maven-toolchains-plugin` or some other strategies resolving to a Maven toolchain, to explicit in your pom that you need Java 8 for your build to work. Then you can configure your toolchains.xml to define where java 8 is located. After, the critical point is to stop using `${java.home}` (which is the Java running Maven, not necessarily the one you want anyway) and find a way to access the configured toolchain location.
That last part is most likely what requires a bit more work. I believe it would be good to request to Maven or to maven-toolchains-plugin a way to access toolchains information from a pom file, with a property such as `${toolchain.jdk.location}` so you could use it in place of `${java.home}`. Maybe there are already some existing workarounds (eg mojos) that take care of exposing toolchains info as properties; but at this stage, it's better to ask on Maven support channels directly (the more general problem is about referencing toolchain locations instead of java.home in pom files).

HTH

**Reply by glhez** [reply]
> The best option seems to get rid of `tools.jar`, so what are you really need it only for compilation as indicated in the linked issue? Then [maven-jdk-tools-wrapper](https://github.com/olivergondza/maven-jdk-tools-wrapper) seems the best option + migrating your code-base to a more modern java version.

I did not test before indicating it, but it does not work - it work in maven, but not in Eclipse, because it resolve the tools.jar against the JDK 17: you don't an error finding the dependency, but you do get compilation errors because the tools.jar was not handled by Eclipse.

You would have to add tools.jar in the Java 8 JRE (as it is not added by default) for the workspace.

This would fix the issue for those, like @demlik  (the original requester) are forced to use tools.jar for whatever reasons (I only wanted to use to play with local JMX url and [VirtualMachine](https://docs.oracle.com/en/java/javase/17/docs/api/jdk.attach/com/sun/tools/attach/VirtualMachine.html))

Another solution would be to isolate the module requiring tools.jar and use an older version of Eclipse, but that really depends on that tools.jar is being used for.
